src/redisearch_rs/rlookup/src/lookup.rs

```diff
@@ -0,0 +1,498 @@
+    /// will deallocate the `ResultProcessor`. This function should only be used by [`RLookup::drop`].
+    ///
+    /// # Safety
+    ///
+    /// 1. The caller must ensure the pointer was previously created through [`Self::into_ptr`].
+    /// 2. The caller has to be careful to never call this method twice for the same pointer, otherwise a
+    ///    double-free or other memory corruptions will occur.
+    /// 3. The caller *must* also ensure that `ptr` continues to be treated as pinned.
+    #[inline]
+    #[cfg_attr(not(test), expect(unused, reason = "used by later stacked PRs"))]
+    unsafe fn from_ptr(ptr: NonNull<Self>) -> Pin<Box<Self>> {
+        // This function must be kept in sync with `Self::into_ptr` above.
+
+        // Safety:
+        // 1 -> This function will only ever be called through `RLookup::drop` below.
+        //      We therefore know - because push_key creates pointers through `into_ptr` - that the invariant is upheld.
+        // 2 -> Has to be upheld by the caller
+        let b = unsafe { Box::from_raw(ptr.as_ptr()) };
+        // Safety: 3 -> Caller has to uphold the pin contract
+        unsafe { Pin::new_unchecked(b) }
+    }
+
+    /// Updates this `RLookupKey`'s fields according to the provided [`ffi::FieldSpec`].
+    pub fn update_from_field_spec(&mut self, fs: &ffi::FieldSpec) {
+        self.flags |= RlookupKeyFlag::DocSrc | RlookupKeyFlag::SchemaSrc;
+
+        let path = {
+            debug_assert!(!fs.fieldPath.is_null());
+
+            let mut path_len = 0;
+            // Safety: we received the pointer from the field spec and have to assume it is valid
+            let path_ptr =
+                unsafe { ffi::HiddenString_GetUnsafe(fs.fieldPath, ptr::from_mut(&mut path_len)) };
+
+            debug_assert!(!path_ptr.is_null());
+            // Safety: we have to assume the pointer and length we get from C here are correct.
+            let bytes = unsafe { slice::from_raw_parts(path_ptr.cast::<u8>(), path_len) };
+            let path = CStr::from_bytes_with_nul(bytes)
+                .expect("string returned by HiddenString_GetUnsafe is malformed");
+
+            if self.flags.contains(RlookupKeyFlag::NameAlloc) {
+                Cow::Owned(path.to_owned())
+            } else {
+                Cow::Borrowed(path)
+            }
+        };
+        self._path = Some(path);
+        self.path = self._path.as_ref().unwrap().as_ptr();
+
+        let fs_options = FieldSpecOptions::from_bits(fs.options()).unwrap();
+
+        if fs_options.contains(FieldSpecOption::Sortable) {
+            self.flags |= RlookupKeyFlag::SvSrc;
+            self.svidx = u16::try_from(fs.sortIdx).unwrap();
+
+            if fs_options.contains(FieldSpecOption::Unf) {
+                // If the field is sortable and not normalized (UNF), the available data in the
+                // sorting vector is the same as the data in the document.
+                self.flags |= RlookupKeyFlag::ValAvailable;
+            }
```

Comment:
The data is the same if it is already normalized, or do I misunderstand?
I think, this is the other way round, see [unf query in c](https://github.com/RediSearch/RediSearch/commit/fa57fd2de377424327afe43f989562210bb226f7#diff-90e39f5f0dfc1344b1b22db34414d3623c2d53cfedcb89e870e24fbb55c93d36L392-R393) and [original string put code](https://github.com/RediSearch/RediSearch/commit/fa57fd2de377424327afe43f989562210bb226f7#diff-2619b7927348152fd98b110f12b1f3683f0a2e13f0f1a793b5b179cd9c704fb7L78-L79)

```suggestion
            if !fs_options.contains(FieldSpecOption::Unf) {
                // If the field is sortable and normalized (not UNF), the available data in the
                // sorting vector is the same as the data in the document.
                self.flags |= RlookupKeyFlag::ValAvailable;
            }
```

Reply:
well, idk 🤷‍♂️ its just what the C code does https://github.com/RediSearch/RediSearch/blob/master/src/rlookup.c#L105

Reply:
The places in C we referenced are contradicting 🤷‍♂️

Reply:
> if (FieldSpec_IsUnf(fs)) {
>       // If the field is sortable and not normalized (UNF), the available data in the
>       // sorting vector is the same as the data in the document.
>       key->flags |= RLOOKUP_F_VAL_AVAILABLE;
>     }